### ISSUE TITLE
Clarify that constrained_layout and tight_layout conflict with each other

### ISSUE THREAD

**Issue by ericpre** [matplotlib/matplotlib#17339]
### Bug report

**Bug summary**

Not sure if this an expected behaviour, but the colorbar seems to be ignored when using `tight_layout`. Example taken from the "Constrained Layout Guide"  (https://matplotlib.org/3.1.3/tutorials/intermediate/constrainedlayout_guide.html#colorbars).

**Code for reproduction**

```python
arr = np.arange(100).reshape((10, 10))
fig, axs = plt.subplots(2, 2, figsize=(4, 4), constrained_layout=True)
for ax in axs.flat:
    im = ax.pcolormesh(arr)
fig.colorbar(im, ax=axs, shrink=0.6)
plt.tight_layout()
```

**Actual outcome**

![image](https://user-images.githubusercontent.com/11851990/81168951-e5663f00-8f8f-11ea-85ea-4dabce14e081.png)

**Expected outcome**
This is the outcome before calling `plt.tight_layout()`

![image](https://user-images.githubusercontent.com/11851990/81167425-185b0380-8f8d-11ea-8010-059d63f93e47.png)

**Matplotlib version**

  * Operating system: Fedora 31
  * Matplotlib version: 3.1.3 from conda defaults (3.2.1 from conda-forge behave similarly)
  * Matplotlib backend (`print(matplotlib.get_backend())`): Qt5Agg
  * Python version: 3.7.6
  * Jupyter version (if applicable):
  * Other libraries:

**Comment by jklymak** [comment]
Yes, thats expected, or at least understood.  You can use `constrained_layout` instead...

**Comment by QuLogic** [comment]
Maybe we should raise if calling `tight_layout` when the figure has `constrained_layout` enabled?

**Comment by jklymak** [comment]
Thats not particularly the problem here; constrained_layout is a bit of a red herring.    `tight_layout` just doesn't know about the big colorbar so ignores it.  In the code above `constrained_layout=True` is ignored because `tight_layout` gets precedence (simply because its older).

**Comment by QuLogic** [comment]
>  In the code above `constrained_layout=True` is ignored because `tight_layout` gets precedence (simply because its older).

Does that generate a warning then?

**Comment by ericpre** [comment]
If `tight_layout` and `constrained_layout` are not compatible, it would useful to have it documented - I can do this at some point.

Regardless of using `constrained_layout` or not, can you confirm the following: if I understand correctly, the colorbar is ignored by `tight_layout` because it is an instances of `matplotlib.axes._axes.Axes`  and not `matplotlib.axes._subplots.AxesSubplot`?

**Comment by jklymak** [comment]
Its says `constrained_layout` is to be used instead of `tight_layout` in quite a few places.  They are not incompatible per se, just tight_layout cancels out constrained_layout.  

We could raise a warning I suppose.

**Comment by tacaswell** [comment]
At least on master we do warn:

```
/tmp/pyIDWTj1:8: UserWarning: This figure includes Axes that are not compatible with tight_layout, so results might be incorrect.
/tmp/pyIDWTj1:8: UserWarning: This figure was using constrained_layout==True, but that is incompatible with subplots_adjust and or tight_layout: setting constrained_layout==False. 
```

This limitation of `tight_layout` was one of the motivations for writing constrained layout.  I think action items here are:

 1. add a sentance to the top indicating that constrained layout and tight layout are alternatives
 2. add a sub-section to https://matplotlib.org/tutorials/intermediate/constrainedlayout_guide.html#limitations showing `tight_layout` breaking this example.

**Comment by ericpre** [comment]
Thanks for the explanations!
In the documentation, it is mentioned that both are alternatives but it is not clear, which one would be more useful. Even if there are many examples, it is not straightforward to work out which one would be more appropriate/robust. Maybe also add some simple guidelines or rule of thumb to the documentation?
